File: solutions/04.enterprise-bot/Dialogs/Shared/EnterpriseDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Luis;
using Microsoft.Bot.Builder.Dialogs;

namespace EnterpriseBot
{
    public class EnterpriseDialog : InterruptableDialog
    {
        // Fields
        private readonly BotServices _services;
        private readonly CancelResponses _responder = new CancelResponses();

        public EnterpriseDialog(BotServices botServices, string dialogId)
            : base(dialogId)
        {
            _services = botServices;

            AddDialog(new CancelDialog());
        }

        protected override async Task<InterruptionStatus> OnDialogInterruptionAsync(DialogContext dc, CancellationToken cancellationToken)
        {
            // Check dispatch intent.
            var luisService = _services.LuisServices["<YOUR MSBOT NAME>_General"];
            var luisResult = await luisService.RecognizeAsync<General>(dc.Context, cancellationToken);
            var intent = luisResult.TopIntent().intent;

            switch (intent)
            {
                case General.Intent.Cancel:
                    {
                        return await OnCancel(dc);
                    }

                case General.Intent.Help:
                    {
                        return await OnHelp(dc);
                    }
            }

            return InterruptionStatus.NoAction;
        }

        protected virtual async Task<InterruptionStatus> OnCancel(DialogContext dc)
        {
            if (dc.ActiveDialog.Id != nameof(CancelDialog))
            {
                // Don't start restart cancel dialog.
                await dc.BeginDialogAsync(nameof(CancelDialog));

                // Signal that the dialog is waiting on user response.
                return InterruptionStatus.Waiting;
            }

            // Else, continue
            return InterruptionStatus.NoAction;
        }

        protected virtual async Task<InterruptionStatus> OnHelp(DialogContext dc)
        {
            var view = new MainResponses();
            await view.ReplyWith(dc.Context, MainResponses.Help);

            // Signal the conversation was interrupted and should immediately continue.
            return InterruptionStatus.Interrupted;
        }
    }
}

File: samples/javascript_typescript/52.enterprise-bot/src/dialogs/shared/enterpriseDialog.ts
import { RecognizerResult } from 'botbuilder';
import { LuisRecognizer } from "botbuilder-ai";
import { DialogContext } from "botbuilder-dialogs";
import { BotServices } from "../../botServices";
import { CancelDialog } from "../cancel/cancelDialog";
import { CancelResponses } from "../cancel/cancelResponses";
import { InterruptableDialog } from "./interruptableDialog";
import { MainResponses } from "../main/mainResponses";
import { TelemetryLuisRecognizer } from "../../middleware/telemetry/telemetryLuisRecognizer";

export class EnterpriseDialog extends InterruptableDialog {

    // Fields
    private readonly _services: BotServices;
    private readonly _cancelResponder: CancelResponses = new CancelResponses();
    private readonly _mainResponder: MainResponses = new MainResponses();

    constructor(botServices: BotServices, dialogId: string) { 
        super(dialogId); 
        
        this._services = botServices;
        this.addDialog(new CancelDialog());
    }
    
    protected async onDialogInterruption(dc: DialogContext): Promise<InterruptionStatus> {
        // Check dispatch intent.
        const luisService: TelemetryLuisRecognizer | undefined = this._services.luisServices.get('<YOUR MSBOT NAME>_General');
        if (!luisService) return Promise.reject(new Error('Luis service not presented'));
        
        const luisResult: RecognizerResult = await luisService.recognize(dc.context);
        const intent: string = LuisRecognizer.topIntent(luisResult);
        
        switch (intent) {
            case 'Cancel':
            return await this.onCancel(dc);
            case 'Help':
            return await this.onHelp(dc);
        }

        return InterruptionStatus.NoAction;
    }

    protected async onCancel(dc: DialogContext): Promise<InterruptionStatus> {
        if (dc.activeDialog && dc.activeDialog.id != 'CancelDialog') {
            // Don't start restart cancel dialog.
            await dc.beginDialog('CancelDialog');

            // Signal that the dialog is waiting on user response.
            return InterruptionStatus.Waiting;
        }

        // Else, continue
        return InterruptionStatus.NoAction;
    }

    protected async onHelp(dc: DialogContext): Promise<InterruptionStatus> {
        this._mainResponder.ReplyWith(dc.context, MainResponses.Help);

        // Signal the conversation was interrupted and should immediately continue.
        return InterruptionStatus.Interrupted;
    }
}
